### ISSUE TITLE
Verb 'ask_for' and Compliance

### ISSUE THREAD

**Issue by tajmone** [AnssiR66/AlanStdLib#18]
While translating the verb `ask_for`, I've noticed the following code ([lines 425–432]):

```alan
      DOES
        MAKE act compliant.
        -- It is only possible to get something from an NPC
        -- if the NPC is 'compliant'.
        LOCATE obj IN hero.
        SAY THE act. "gives" SAY THE obj. "to you."
        MAKE act NOT compliant.

```

It seems that this verb doesn't use a temporary attribute to store the state of the `ACTOR`'s compliance before its executiion, which will result in an `ACTOR` always becoming compliant after its execution.

Isn't this a potential source of problems, because of compliant actors becoming non compliant through the use of the verb? Shouldn't this verb store and restore the `ACTOR`'s state of the `compliant` attribute?

Also, the message should probably check if `act` is plural, and print "give" if it is:

```alan
        SAY THE act. "gives" SAY THE obj. "to you."
```

I could fix both issues, but I'd rather wait for a confirmation from you, at least for the first one.

[lines 425–432]: https://github.com/AnssiR66/AlanStdLib/blob/741055b/lib_verbs.i#L425

**Comment by AnssiR66** [comment]
Hi Tristano!
Yes, good catch! Feel free to fix both issues please, I will merge them once committed.

Thanks,

Anssi


________________________________

**Comment by tajmone** [comment]
I've added a test file in commit 335215e to help me keep track of this problem while I apply the changes.

So I'll need to apply to fixes to the new "test-folder" branch of PR #17 in order to check that the test passes correctly after the fix. The test-folder branch hasn't been merged, but in any case the single commit with the fix could then be cherry-picked into master, or I could make a new branch and cherry pick it there, and make a PR for the fix commit only.

I know that I haven't been able to add as many tests as I wished to, but I think that having a tests folder is a good idea as it really helps to catch any unexpected side-effects when tweaking the library, as well as to catch some minor existing problems.

Anyhow, I'll start working on the fix and then be back to you...

**Comment by tajmone** [comment]
OK, the problem was fixed with commit 1bf124a!

I've also changed the compliance test (commit 31b9dc0d) to check that the verb `ask_for` now preserves the original actor's state with both a compliant and a non-compliant actor. 

As you can see from the [new test transcript], now all tests pass:

```
> ;-----------------------
> ; Woman compliance-test:
> ;-----------------------
> test compliance of woman
The woman is compliant.

> ask woman for mirror
The woman gives the magic mirror to you.

> test compliance of woman
The woman is compliant.

> ; OK! compliance is as before.
> ;---------------------
> ; Man compliance-test:
> ;---------------------
> test compliance of man
The man is not compliant.

> ask man for sword
The man gives the sword to you.

> test compliance of man
The man is not compliant.

> ; OK! compliance is as before.
```

As mentioned before, for practical reasons the fix-commit was made on the branch of the "test-folder" PR (which is still unmerged). Can you merge the whole PR, or do I need to make a separate PR with just the verb fix?


[new test transcript]:https://github.com/tajmone/AlanStdLib/blob/31b9dc0d/tests/actors-compliance.a3log#L29
